cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -269,29 +285,406 @@ func copyFromGlobHelper(src, dst, globEx string) error {
+	disk disks.Disk
+
+	isEncryptedDev bool
+
+	// state for tracking what happens as we progress through degraded mode of
+	// recovery
+	degradedState *recoverDegradedState
+}
+
+// degraded returns whether a degraded recover mode state has fallen back from
+// the typical operation to some sort of degraded mode.
+func (m *stateMachine) degraded() bool {
+	r := m.degradedState
+
+	if m.isEncryptedDev {
+		// for encrypted devices, we need to have ubuntu-save mounted
+		if r.UbuntuSave.MountState != partitionMounted {
+			return true
+		}
+
+		// we also should have all the unlock keys as run keys
+		if r.UbuntuData.UnlockKey != keyRun {
+			return true
+		}
+
+		if r.UbuntuSave.UnlockKey != keyRun {
+			return true
+		}
+	} else {
+		// for unencrypted devices, ubuntu-save must either be mounted or
+		// absent-but-optional
+		if r.UbuntuSave.MountState != partitionMounted {
+			if r.UbuntuSave.MountState != partitionAbsentOptional {
+				return true
+			}
+		}
+	}
+
+	// ubuntu-boot and ubuntu-data should both be mounted
+	if r.UbuntuBoot.MountState != partitionMounted {
+		return true
+	}
+	if r.UbuntuData.MountState != partitionMounted {
+		return true
+	}
+
+	// TODO: should we also check MountLocation too?
+
+	// we should have nothing in the error log
+	if len(r.ErrorLog) != 0 {
+		return true
+	}
+
+	return false
+}
+
+func (m *stateMachine) diskOpts() *disks.Options {
+	if m.isEncryptedDev {
+		return &disks.Options{
+			IsDecryptedDevice: true,
+		}
+	}
+	return nil
+}
+
+func (m *stateMachine) verifyMountPoint(dir, name string) error {
+	matches, err := m.disk.MountPointIsFromDisk(dir, m.diskOpts())
+	if err != nil {
 		return err
 	}
+	if !matches {
+		return fmt.Errorf("cannot validate mount: %s mountpoint target %s is expected to be from disk %s but is not", name, dir, m.disk.Dev())
+	}
+	return nil
+}
 
-	// get the disk that we mounted the ubuntu-seed partition from as a
-	// reference point for future mounts
-	disk, err := disks.DiskFromMountPoint(boot.InitramfsUbuntuSeedDir, nil)
+func (m *stateMachine) setFindState(part, partUUID string, err error, logNotFoundErr bool) error {
+	if err == nil {
+		// device was found
+		part := m.degradedState.partition(part)
+		part.FindState = partitionFound
+		part.Device = fmt.Sprintf("/dev/disk/by-partuuid/%s", partUUID)
+		return nil
+	}
+	if _, ok := err.(disks.FilesystemLabelNotFoundError); ok {
+		// explicit error that the device was not found
+		m.degradedState.partition(part).FindState = partitionNotFound
+		if logNotFoundErr {
+			m.degradedState.LogErrorf("cannot find %v partition on disk %s", part, m.disk.Dev())
+		}
+		return nil
+	}
+	// the error is not "not-found", so we have a real error
+	m.degradedState.partition(part).FindState = partitionErrFinding
+	m.degradedState.LogErrorf("error finding %v partition on disk %s: %v", part, m.disk.Dev(), err)
+	return nil
+}
+
+func (m *stateMachine) setMountState(part, where string, err error) error {
 	if err != nil {
```

Comment:
(nitpick^2) I wonder if it would be worthwhile to try to keep the set*State similar in structure, i.e. start with "err == nil" or "err != nil" in all of them first etc. But maybe they are too different for this to be worthwhile.

Reply:
done in https://github.com/snapcore/snapd/pull/9621